src/fparser/two/Fortran2008.py

```diff
@@ -1341,6 +1350,234 @@ def match(string):
+        "Execution_Part_Construct",
+        "End_Block_Stmt",
+    ]
+
+    @staticmethod
+    def match(reader):
+        return BlockBase.match(
+            Block_Stmt,
+            [Specification_Part, Execution_Part_Construct],
+            End_Block_Stmt,
+            reader,
+            match_names=True,  # C810
+            strict_match_names=True,  # C810
+        )
+
+
+class Block_Stmt(StmtBase, WORDClsBase, ScopingRegionMixin):
+    """
+    Fortran 2008 Rule 808.
+
+    block-stmt is [ block-construct-name : ] BLOCK
+
+    """
+
+    subclass_names = []
+    use_names = ["Block_Construct_Name"]
+    counter = 0
+
+    @staticmethod
+    def match(string):
+        """
+        Attempts to match the supplied text with this rule.
+
+        :param str string: the text to match.
+
+        :returns: a tuple of the matched node and instance of Counter or \
+                  None if there is no match.
+        :rtype: Tuple["BLOCK", \
+                      :py:class:`fparser.two.Fortran2008.Block_Stmt.Counter`] \
+                | NoneType
+        """
+        found = WORDClsBase.match("BLOCK", None, string)
+        if not found:
+            return None
+        block, _ = found
+        # Construct a unique name for this BLOCK (in case it isn't named). We
+        # ensure the name is not a valid Fortran name so that it can't clash
+        # with any regions named in the code.
+        scope_name = f"block:{Block_Stmt.counter}"
+        Block_Stmt.counter += 1
+        # TODO #397. Ideally we'd have the name associated with the Block
+        # Construct here (if any) so that it could be displayed in repr.
+        # As it is, repr will show scope_name which will not be the same
+        # as any explicit name given to the Block. (This name *is* shown
+        # in the repr of the End_Block_Stmt.) This problem is common to
+        # other block constructs such as Block_Nonlabel_Do_Construct.
+        return block, scope_name
+
+    def get_scope_name(self):
+        """
+        :returns: the name of this scoping region.
+        :rtype: str
+        """
+        if self.item.name:
+            return self.item.name
+        return self.items[1]
+
+    def get_start_name(self):
+        """
+        :returns: the name associated with this Block construct or None.
+        :rtype: str | NoneType
+        """
+        return self.item.name
+
+    def tostr(self):
+        """
+        :returns: the string representation of this node.
+        :rtype: str
+        """
+        return "BLOCK"
+
+
+class End_Block_Stmt(EndStmtBase):  # R809
+    """
+    Fortran 2008 Rule 809.
+
+    end-block-stmt is END BLOCK [ block-construct-name ]
+
+    """
+
+    subclass_names = []
+    use_names = ["Block_Construct_Name"]
+
+    @staticmethod
+    def match(string):
+        """
+        :param str string: Fortran code to check for a match
+
+        :return: code line matching the "END BLOCK" statement
+        :rtype: str
```

Comment:
Here and in End_Critical_Stmt the rtype is str which doesn't seem right. Looking at the EndStmtBase class it seems to return "Tuple[stmt_type, str? | NoneType] | NoneType" but I am not sure what stmt_type is.